### ISSUE TITLE
Any option to use tools on let's say last stage to save the response or parse relevant information to DB

### ISSUE THREAD

**Issue by ranjandsingh** [filip-michalsky/SalesGPT#10]

**Comment by filip-michalsky** [comment]
@ranjandsingh yes, it's been on the TODO list for a while. I will try to push out as early as I can. If you would be open to contribute, please let me know :)

**Comment by Bilal-irfan** [comment]
@filip-michalsky which API or tool you used for cold calling.
when you will update the codebase demo you shared?

**Comment by ranjandsingh** [comment]
@filip-michalsky I'm down to add this option but there are few ways we can implement it 
- Adding another stage and calling save on last stage
- or adding an Agent instead of chain for second chain
- or even in the post process of the conversation like end conversation and detecting if user intend to save

**Comment by filip-michalsky** [comment]
@Bilal-irfan thanks for the question - I am going to push an example with calling an API I am building which complements SalesGPT agents for cold-calling soon. Will need to start a beta testers list as I can't support high traffic yet. Would love to get your feedback.

@ranjandsingh thanks for your offer to help! Adding tools to the AI Sales Agent toolkit is a high priority item, I just haven't had time to work on it yet. What do you mean by the intent to save? 

Rough API sketch (idea, subject to feedback):

1. User input
2. Stage analyzer outputs which stage of the sales conversation we are at.
(optional) - new config, we init the sales agent with a different prompt based on the stage. Right now, the sales agent has a static prompt.
3. Sales Agent decides what to do. It can look up product information in a knowledge base. It can send an email, or text, it can initiate a phone call, search the internet, and output a response to the user.
4. Agent Critic - does the output match the stage and does it match the product knowledge base? if not, give feedback and re-try
5. User output.

**Comment by tgbaoo** [comment]
Hi filip, really appreciate your work, the only problem of this is turn the salegpt become an independent agents - plan and execute (independenly from human on decide what to do), from What I see this is just an action agents or a simple chain - from langchain, so i have an idea that we add the conversation analyzer and conversation chain become a part of tools to a plan and execute chain. So could we start this discuss about mine idea and discuss how we implement and try it as a raw version on python?

**Comment by Alex-Poon** [comment]
@filip-michalsky I'm learning how to use tools by adding tools to the Sales Agent. I see in the examples they call `initialize_agent` which returns an AgentExecutor. Would this work by returning an AgentExecutor from the `from_llm` method?

**Comment by thalex** [comment]
Hi Filip, I am testing your salesgpt code and have a doubt. Where do I add the number to receive the call and do I have to set a call duration time?

**Comment by filip-michalsky** [comment]
@Alex-Poon @thalex @ranjandsingh @Bilal-irfan @tgbaoo Agent example is up, check it out! Will close this issue soon if you don't have any objections.

**Comment by ranjandsingh** [comment]
well, this is a good option and value add to the current System love to see it.  just one question does it supports multiple tools or just one ?

**Comment by filip-michalsky** [comment]
@ranjandsingh yessir, you can add as many tools as you want, you have to just change the get_tools() function

**Comment by filip-michalsky** [comment]
I guess you have a good point, it's not now easily changeable by the user since it's not configurable from the high-level API.

We can add it as a TODO to allow a user to specify their own tools.

**Comment by avyaymc** [comment]
Does anyone have a working demo of multi-tools? seems to me its very bad at telling the user what tools it has available, any ideas on how to modify the prompt to get that capacity?

**Comment by ranjandsingh** [comment]
@avyaymc It's on todo for now you might need to modify the agent code to allow multiple tools currently only accepting one

**Comment by filip-michalsky** [comment]
@avyaymc adding this to TODO - open for someone to contribute

**Comment by Eran-BA** [comment]
any news?

**Comment by tgbaoo** [comment]
Hi, I am seeing that the code is currently not working as an langchain agent with tools right now? If yes, i am developing some structure for that, hope to contribute to this repo

**Comment by tgbaoo** [comment]
Hi, I am seeing that the code is currently not working as an langchain
agent with tools right now? If yes, i am developing some structure for
that, hope to contribute to this repo

Vào Th 6, 6 thg 10, 2023 vào lúc 16:14 Eran Ben Artzy <
***@***.***> đã viết:

**Comment by tgbaoo** [comment]
By the way, we also need to test the benchmark of the retrieval document due to specific document's field and also control the way the retrieval come to the prompt to generate the answer by the langchain agent also right? @filip-michalsky
